File: windows_csharp/RimeoAgent/HttpServer/SecurityGates.cs
using System.Security.Cryptography;
using System.Text;
using RimeoAgent.Config;
using RimeoAgent.Services;

namespace RimeoAgent.HttpServer;

// Fail-CLOSED security primitives — port of macOS SecurityGates.swift. Covers the
// agent security audit:
//   • 6001  jwtGate fail-open   → AccessControl.Decide (deny when no tunnel + no PSK)
//   • 6004  control endpoints   → AccessControl.ControlProtectedPaths
//   • 6003  CORS "*"            → CorsPolicy (explicit origin allow-list)
//   • 6002  arbitrary file read → LibraryPathGuard (canonicalize + root allow-list)
//   • 6005  unsigned auto-update → UpdateSignatureVerifier (detached ES256 over the zip)
//   • 6006  helper download     → ComponentHostPolicy (domain pinning)

public enum AccessDecision { Allow, Deny }

public static class AccessControl
{
    // Data-exposing endpoints (audio bytes, waveform, art, library JSON, logs).
    public static readonly HashSet<string> DataProtectedPaths = new()
    {
        "/stream", "/waveform", "/artwork", "/api/data", "/api/logs",
    };

    // Mutating / control endpoints. Unauthenticated = account takeover, forced
    // public exposure, arbitrary explorer /select, note tampering (task 6004).
    public static readonly HashSet<string> ControlProtectedPaths = new()
    {
        "/reveal",
        "/api/save_note", "/api/save_exclusions", "/api/rename_history",
        "/api/send_tg",
        "/api/link_account", "/api/unlink_account",
        "/api/agent_login", "/api/agent_signup",
        "/api/tunnel/start", "/api/tunnel/stop",
        "/api/report_bug",
        // Мутации плейлистов (Фаза 0 — плейлисты из iOS). Без авторизации любой в той же
        // сети переписывает оверлей чужой библиотеки. /api/playlist/recommendations
        // остаётся ПУБЛИЧНЫМ (как /api/similar) — он только читает.
        "/api/playlist/create", "/api/playlist/create_folder",
        "/api/playlist/add", "/api/playlist/remove",
        "/api/playlist/reorder", "/api/playlist/rename", "/api/playlist/delete",
        // Фаза 6 — единственный роут, который пишет в НАСТОЯЩИЙ master.db Rekordbox
        // (создание/переименование/удаление плейлистов, перезапись состава). Все роуты
        // выше правят наш собственный оверлей, который не жалко выбросить; этот —
        // незаменимую библиотеку. Неаутентифицированным POST'ом сюда любой сосед по
        // Wi-Fi сносит диджею плейлисты, поэтому он в контрольной корзине, а не в
        // «чтении» (риск 14). /sync/status тоже защищён — паритет с macOS
        // (controlProtectedPaths), хотя это GET: он раскрывает состав очереди синка.
        "/api/playlist/sync",
        "/api/playlist/sync/status",
        // Самообновление по запросу с телефона: ставит новый бинарь и перезапускает
        // агент — без авторизации это удалённая подмена исполняемого файла.
        "/api/agent/update", "/api/agent/update/status",
    };

    public static bool RequiresAuth(string path) =>
        DataProtectedPaths.Contains(path) || ControlProtectedPaths.Contains(path);

    /// Pure authorization core. Order: valid PSK → allow; else no named tunnel ⇒
    /// DENY (6001 fail-closed — was allow); else allow iff the server JWT validates.
    public static AccessDecision Decide(
        string lanSecret, string? providedToken, string namedHostname,
        string? jwtToken, Func<string?, string, JwtValidator.Failure?> validate)
    {
        if (!string.IsNullOrEmpty(lanSecret) && !string.IsNullOrEmpty(providedToken)
            && ConstantTimeEquals(providedToken!, lanSecret))
            return AccessDecision.Allow;
        if (string.IsNullOrEmpty(namedHostname)) return AccessDecision.Deny; // 6001
        return validate(jwtToken, namedHostname) == null
            ? AccessDecision.Allow : AccessDecision.Deny;
    }

    /// Constant-time PSK comparison (no length/prefix timing leak).
    public static bool ConstantTimeEquals(string a, string b) =>
        CryptographicOperations.FixedTimeEquals(
            Encoding.UTF8.GetBytes(a), Encoding.UTF8.GetBytes(b));
}

public static class CorsPolicy
{
    // Only the Rimeo web player talks to the agent cross-origin. Native iOS / relay
    // clients are not browsers and need no CORS. A wildcard "*" let ANY page fetch
    // /api/data and /stream (task 6003).
    public static readonly HashSet<string> AllowedOrigins = new()
    {
        "https://rimeo.app", "https://www.rimeo.app",
    };

    public static bool IsAllowed(string? origin) =>
        !string.IsNullOrEmpty(origin) && AllowedOrigins.Contains(origin!);

    public static Dictionary<string, string> Headers(string? origin)
    {
        if (!IsAllowed(origin)) return new Dictionary<string, string>();
        return new Dictionary<string, string>
        {
            ["Access-Control-Allow-Origin"]  = origin!,
            ["Access-Control-Allow-Methods"] = "GET, POST, OPTIONS",
            ["Access-Control-Allow-Headers"] = "Authorization, Content-Type, Range",
            ["Vary"]                         = "Origin",
        };
    }
}

public static class LibraryPathGuard
{
    private static readonly object RootsLock = new();
    private static string _cachedRootsSignature = "";
    private static List<string> _cachedCanonicalRoots = new();

    /// Канонический абсолютный путь: убрать `..` (GetFullPath) и разрешить
    /// symlink/junction в конечную цель, чтобы ссылка изнутри библиотеки не уводила
    /// наружу.
    ///
    /// ⚠️ РЕЗОЛВИМ КАЖДЫЙ КОМПОНЕНТ, А НЕ ТОЛЬКО ПОСЛЕДНИЙ. Swift зовёт realpath(),
    /// который разворачивает ссылки на ВСЕХ уровнях пути. Прошлая версия дёргала
    /// ResolveLinkTarget только у листа — и этого достаточно, чтобы сбежать из
    /// контейнмента через junction в СЕРЕДИНЕ пути:
    ///     mklink /J D:\Music\esc C:\Users\<u>
    ///     ?path=D:\Music\esc\.ssh\id_rsa
    /// Лист (.ssh\id_rsa) ссылкой не является, ResolveLinkTarget вернул бы null,
    /// путь остался бы «D:\Music\esc\.ssh\id_rsa» — формально внутри корня D:\Music,
    /// а физически это C:\Users\<u>\.ssh\id_rsa. Регрессия 6002.
    public static string Canonical(string path)
    {
        try
        {
            var full = Path.GetFullPath(path);
            var root = Path.GetPathRoot(full) ?? "";
            if (root.Length == 0) return full;

            var rest  = full.Substring(root.Length);
            var parts = rest.Split(new[] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar },
                                   StringSplitOptions.RemoveEmptyEntries);

            var cur = root;
            foreach (var part in parts)
            {
                cur = Path.Combine(cur, part);
                try
                {
                    // returnFinalTarget: true разворачивает и цепочку ссылок целиком.
                    // На не-ссылке вернёт null; на несуществующем пути (том не
                    // примонтирован) — бросит или вернёт null, и мы остаёмся на
                    // лексической форме, как и Swift в своём фолбэке.
                    FileSystemInfo? target = Directory.Exists(cur)
                        ? new DirectoryInfo(cur).ResolveLinkTarget(returnFinalTarget: true)
                        : new FileInfo(cur).ResolveLinkTarget(returnFinalTarget: true);
                    if (target != null) cur = Path.GetFullPath(target.FullName);
                }
                catch { /* не ссылка / нет доступа / не существует — берём как есть */ }
            }
            return Path.GetFullPath(cur);
        }
        catch { return path; }
    }

    /// Directory-prefix containment over already-canonical inputs (case-insensitive;
    /// Windows FS), with a separator boundary so root "D:\Music" does NOT authorise
    /// "D:\Music-secret".
    public static bool IsContainedCanonical(string candidate, IEnumerable<string> canonicalRoots)
    {
        if (string.IsNullOrEmpty(candidate)) return false;
        foreach (var r in canonicalRoots)
        {
            if (string.IsNullOrEmpty(r)) continue;
            if (string.Equals(candidate, r, StringComparison.OrdinalIgnoreCase)) return true;
            var boundary = r.EndsWith(Path.DirectorySeparatorChar) ? r : r + Path.DirectorySeparatorChar;
            if (candidate.StartsWith(boundary, StringComparison.OrdinalIgnoreCase)) return true;
        }
        return false;
    }

    public static bool IsContained(string path, IEnumerable<string> roots)
        => IsContainedCanonical(Canonical(path), roots.Select(Canonical));

    /// Корень СИСТЕМНОГО тома (обычно "C:\") — прямой аналог "/" на macOS, который
    /// Swift отбрасывает явно (`dir != "/"`).
    ///
    /// ⚠️ ЗАЧЕМ. Трек, лежащий ПРЯМО в корне системного диска ("C:\song.mp3"), даёт
    /// корень "C:\" — и тогда IsContainedCanonical пропускает ЛЮБОЙ файл диска:
    /// ?path=C:\Users\<u>\.ssh\id_rsa проходит через /stream, /waveform, /artwork.
    /// Это ровно та дыра, которую закрывала задача 6002.
    ///
    /// Корни ДРУГИХ томов ("D:\", UNC-шара) НЕ отбрасываем осознанно: это аналог
    /// /Volumes/KODAK, который macOS-агент разрешает. Иначе выделенный музыкальный
    /// диск, где треки лежат прямо в корне, перестал бы играть вовсе — а системных
    /// секретов там нет.
    private static bool IsSystemVolumeRoot(string dir)
    {
        var root = Path.GetPathRoot(dir);
        if (string.IsNullOrEmpty(root)) return false;
        var trimmed = dir.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        var rootTrimmed = root.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        if (!string.Equals(trimmed, rootTrimmed, StringComparison.OrdinalIgnoreCase)) return false;

        var windows = Environment.GetFolderPath(Environment.SpecialFolder.Windows);
        var sysRoot = Path.GetPathRoot(windows);
        if (string.IsNullOrEmpty(sysRoot)) return false;
        return string.Equals(rootTrimmed,
                             sysRoot.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar),
                             StringComparison.OrdinalIgnoreCase);
    }

    /// Parent folder of every library track + the agent's cache (converted audio)
    /// and the Rekordbox share/db dir (cover art).
    public static List<string> AllowedRoots()
    {
        var roots = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var t in RekordboxParser.Shared.Parse().Tracks)
        {
            if (string.IsNullOrEmpty(t.Location)) continue;
            var dir = Path.GetDirectoryName(t.Location);
            if (string.IsNullOrEmpty(dir) || IsSystemVolumeRoot(dir!)) continue;
            roots.Add(dir!);
        }
        if (!string.IsNullOrEmpty(AppConfig.Shared.CacheDir)) roots.Add(AppConfig.Shared.CacheDir);
        var db = AppConfig.Shared.DbPath;
        if (!string.IsNullOrEmpty(db))
        {
            var rb = Path.GetDirectoryName(db);
            if (!string.IsNullOrEmpty(rb) && !IsSystemVolumeRoot(rb!))
            {
                roots.Add(rb!);
                roots.Add(Path.Combine(rb!, "share"));
            }
        }
        return roots.ToList();
    }

    /// Canonicalized roots, memoized on a cheap library signature so the hot
    /// streaming path does not canonicalize every root on every (range) request.
    private static List<string> CanonicalAllowedRoots()
    {
        var tracks = RekordboxParser.Shared.Parse().Tracks;
        var signature = $"{tracks.Count}|{(tracks.Count > 0 ? tracks[0].Location : "")}|{(tracks.Count > 0 ? tracks[^1].Location : "")}";
        lock (RootsLock)
        {
            if (signature == _cachedRootsSignature && _cachedCanonicalRoots.Count > 0)
                return _cachedCanonicalRoots;
            _cachedCanonicalRoots = AllowedRoots().Select(Canonical).ToList();
            _cachedRootsSignature = signature;
            return _cachedCanonicalRoots;
        }
    }

    public static bool IsAllowed(string path) =>
        !string.IsNullOrEmpty(path) && IsContainedCanonical(Canonical(path), CanonicalAllowedRoots());
}

public sealed class UpdateSignatureException : Exception
{
    public UpdateSignatureException(string message) : base(message) { }
}

public static class UpdateSignatureVerifier
{
    // Contents of update_pub.pem (PKIX/SPKI). Matching private key lives ONLY in
    // release CI (secret UPDATE_SIGNING_KEY) and signs update.zip → update.zip.sig.
    // Same key as the macOS agent, so one signature verifies on both platforms.
    public const string UpdatePublicKeyPem =
        "-----BEGIN PUBLIC KEY-----\n" +
        "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE13+tB8Bq9f/Hi7ez4nff0roiRz4E\n" +
        "LGxTQT37N0MMd+jwfqh4P+vmK4oMBsrvwhxkIRB60rgWcVbr8irPMYJLEw==\n" +
        "-----END PUBLIC KEY-----";

    /// Verify an update archive against its detached ECDSA-P256 / SHA-256 (DER)
    /// signature — produced by `openssl dgst -sha256 -sign update_priv.pem`.
    /// Throws (fail-CLOSED) when the .sig is absent or does not verify; the caller
    /// MUST NOT apply the update on throw.
    public static void VerifyZipSignature(string zipPath, string sigPath)
        => VerifyZipSignature(zipPath, sigPath, UpdatePublicKeyPem);

    public static void VerifyZipSignature(string zipPath, string sigPath, string publicKeyPem)
    {
        if (!File.Exists(sigPath))
            throw new UpdateSignatureException($"Update signature file missing: {sigPath}");

        var sig = File.ReadAllBytes(sigPath);
        byte[] hash;
        using (var sha = SHA256.Create())
        using (var fs = File.OpenRead(zipPath))
            hash = sha.ComputeHash(fs);

        using var ecdsa = ECDsa.Create();
        try { ecdsa.ImportFromPem(publicKeyPem); }
        catch (Exception ex) { throw new UpdateSignatureException($"Bad update public key: {ex.Message}"); }

        bool ok;
        try { ok = ecdsa.VerifyHash(hash, sig, DSASignatureFormat.Rfc3279DerSequence); }
        catch { ok = false; }
        if (!ok) throw new UpdateSignatureException("Update archive signature is invalid");
    }
}

public static class ComponentHostPolicy
{
    // Hosts we accept helper-binary (cloudflared / ffmpeg / ffprobe) downloads
    // from. Pinning the host means a compromised manifest cannot redirect the
    // download to an attacker server to drop an executable (task 6006).
    public static readonly HashSet<string> AllowedHosts = new(StringComparer.OrdinalIgnoreCase)
    {
        "rimeo.app", "www.rimeo.app",
        "github.com",
        "objects.githubusercontent.com",
        "release-assets.githubusercontent.com",
    };

    public static bool IsAllowed(string urlString)
    {
        if (!Uri.TryCreate(urlString, UriKind.Absolute, out var uri)) return false;
        if (!string.Equals(uri.Scheme, "https", StringComparison.OrdinalIgnoreCase)) return false;
        return AllowedHosts.Contains(uri.Host);
    }
}

File: macos_arm64/Sources/RimeoAgentMac/HTTPServer/SecurityGates.swift
import Foundation
import CryptoKit

// Security hardening primitives shared by the HTTP layer and the updater.
// Every decision here is fail-CLOSED: an unknown/misconfigured state denies
// access rather than granting it. The pure functions (no singleton access) are
// unit-tested in Tests/RimeoAgentTests/SecurityGatesTests.swift.
//
// Covers the agent security audit:
//   • 6001  jwtGate fail-open      → AccessControl.decide (deny when no tunnel + no PSK)
//   • 6004  control endpoints      → AccessControl.controlProtectedPaths
//   • 6003  CORS "*"               → CORSPolicy (explicit origin allow-list)
//   • 6002  arbitrary file read    → LibraryPathGuard (canonicalize + root allow-list)
//   • 6005  unsigned auto-update   → UpdateSignatureVerifier (Team ID + codesign)
//   • 6006  helper-binary download → ComponentHostPolicy (domain pinning)

// MARK: - Access control (6001 fail-closed gate, 6004 control-endpoint auth)

enum AccessDecision: Equatable {
    case allow
    case deny
}

enum AccessControl {
    /// Endpoints that expose user data (audio bytes, waveform, art, the full
    /// library JSON, logs) over the public named tunnel.
    static let dataProtectedPaths: Set<String> = [
        "/stream", "/waveform", "/artwork", "/api/data", "/api/logs",
    ]

    /// Mutating / control endpoints. Left unauthenticated these enable account
    /// takeover (/api/link_account), forced public exposure (/api/tunnel/start),
    /// arbitrary `open -R` (/reveal) and note/exclusion tampering (task 6004).
    static let controlProtectedPaths: Set<String> = [
        "/reveal",
        "/api/save_note", "/api/save_exclusions", "/api/rename_history",
        "/api/send_tg",
        "/api/link_account", "/api/unlink_account",
        "/api/agent_login", "/api/agent_signup",
        "/api/tunnel/start", "/api/tunnel/stop",
        "/api/report_bug",
        // Playlist mutations (Фаза 0 — плейлисты из iOS). Unauthenticated writes
        // would let anyone tamper with a user's library overlay. Kept as a single
        // list; /api/playlist/recommendations stays PUBLIC (like /api/similar).
        "/api/playlist/create", "/api/playlist/add", "/api/playlist/remove",
        "/api/playlist/reorder", "/api/playlist/rename", "/api/playlist/delete",
        "/api/playlist/create_folder",
        // Фаза 6 — the ONLY endpoint that writes the user's real Rekordbox
        // master.db (create/rename/delete playlists, rewrite membership). Every
        // route above it edits a Rimeo-owned overlay file we can throw away; this
        // one mutates the irreplaceable library. Unauthenticated, anyone on the
        // same café/hotel Wi-Fi could POST it and wipe a DJ's playlists — so it is
        // gated exactly like the other mutations, never as a "read" (риск 14).
        "/api/playlist/sync",
        "/api/playlist/sync/status",
        // Обновление агента по запросу с телефона. Неаутентифицированный POST сюда —
        // это «любой в кафейном Wi-Fi может в любой момент перезапустить твой агент
        // посреди сета» (и заставить его качать zip). Подпись обновления это не
        // защищает: она гарантирует, ЧТО поставится, но не КТО это инициировал.
        "/api/agent/update",
        "/api/agent/update/status",
    ]

    /// True when a request to `path` must be authorised (PSK, JWT, or trusted
    /// in-process origin). The router skips the gate only for public endpoints
    /// (pairing handshake, status, similar, analysis reads).
    static func requiresAuth(path: String) -> Bool {
        dataProtectedPaths.contains(path) || controlProtectedPaths.contains(path)
    }

    /// Pure authorization core shared by the live `authGate`. Order:
    ///  1. A valid per-device PSK authorises a local (LAN) client.
    ///  2. No named tunnel (audience empty) ⇒ **DENY** (task 6001 fix). Without a
    ///     named tunnel the server signs no JWT, so the PSK is the only trusted
    ///     remote credential; a request lacking it must be refused, never waved
    ///     through. This is the single line that turns fail-open into fail-closed.
    ///  3. On a named tunnel, allow iff the server JWT validates.
    static func decide(
        lanSecret: String,
        providedToken: String?,
        namedHostname: String,
        jwtToken: String?,
        validate: (_ token: String?, _ audience: String) -> JWTValidator.Failure?
    ) -> AccessDecision {
        if !lanSecret.isEmpty, let provided = providedToken, !provided.isEmpty,
           constantTimeEquals(provided, lanSecret) {
            return .allow
        }
        // 6001: fail closed. Was `return .allow` (open) when the audience was empty.
        guard !namedHostname.isEmpty else { return .deny }
        return validate(jwtToken, namedHostname) == nil ? .allow : .deny
    }

    /// Constant-time comparison so PSK verification does not leak length/prefix via
    /// response timing. Returns true only when both byte sequences are identical.
    static func constantTimeEquals(_ a: String, _ b: String) -> Bool {
        let ab = Array(a.utf8)
        let bb = Array(b.utf8)
        var diff = UInt8(ab.count == bb.count ? 0 : 1)
        let n = max(ab.count, bb.count)
        var i = 0
        while i < n {
            let x = i < ab.count ? ab[i] : 0
            let y = i < bb.count ? bb[i] : 0
            diff |= (x ^ y)
            i += 1
        }
        return diff == 0
    }
}

// MARK: - CORS (6003: no wildcard Access-Control-Allow-Origin)

enum CORSPolicy {
    /// Browsers that legitimately talk to the agent cross-origin: only the Rimeo
    /// web player. Native iOS / relay clients are not browsers and need no CORS at
    /// all. A wildcard "*" let ANY page the user opened fetch /api/data & /stream.
    static let allowedOrigins: Set<String> = [
        "https://rimeo.app",
        "https://www.rimeo.app",
    ]

    static func isAllowed(origin: String?) -> Bool {
        guard let origin = origin, !origin.isEmpty else { return false }
        return allowedOrigins.contains(origin)
    }

    /// CORS response headers for a given request Origin. Empty when the origin is
    /// not on the allow-list → the browser blocks the cross-origin read AND the
    /// pre-flight for a state-changing request fails, so it is never sent.
    static func headers(forOrigin origin: String?) -> [String: String] {
        guard isAllowed(origin: origin), let origin = origin else { return [:] }
        return [
            "Access-Control-Allow-Origin":  origin,
            "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
            "Access-Control-Allow-Headers": "Authorization, Content-Type, Range",
            "Vary":                         "Origin",
        ]
    }
}

// MARK: - Library path containment (6002: no arbitrary file read via ?path=)

enum LibraryPathGuard {
    private static let rootsLock = NSLock()
    private static var cachedRootsSignature = ""
    private static var cachedCanonicalRoots: [String] = []

    /// Canonical absolute path: resolve symlinks + `..` via realpath when the file
    /// exists, else fall back to a lexical standardization so a path on a
    /// not-yet-mounted volume is still stripped of `..` traversal.
    static func canonical(_ path: String) -> String {
        if let real = realpathResolve(path) { return real }
        return URL(fileURLWithPath: path).standardizedFileURL.path
    }

    private static func realpathResolve(_ path: String) -> String? {
        path.withCString { cstr -> String? in
            guard let resolved = realpath(cstr, nil) else { return nil }
            defer { free(resolved) }
            return String(cString: resolved)
        }
    }

    /// Collapse a leading `/Users/<name>/` to `/Users/*/` so username drift between
    /// the exported library and the current macOS account cannot defeat
    /// containment. The resolver in APIRouter deliberately remaps
    /// `/Users/<old>/…` → the current $HOME, so roots and candidates must compare
    /// on a user-agnostic form.
    static func normalizeUser(_ path: String) -> String {
        let marker = "/Users/"
        guard path.hasPrefix(marker) else { return path }
        let rest = path.dropFirst(marker.count)
        guard let slash = rest.firstIndex(of: "/") else { return path }
        return "/Users/*/" + rest[rest.index(after: slash)...]
    }

    /// Pure containment over already-canonical inputs: is `candidate` inside one of
    /// `canonicalRoots`? Directory-prefix with a `/` boundary so root `/a/music`
    /// does NOT authorise `/a/music-secret`.
    static func isContainedCanonical(_ candidate: String, canonicalRoots: [String]) -> Bool {
        guard !candidate.isEmpty else { return false }
        for r in canonicalRoots {
            guard !r.isEmpty else { continue }
            if candidate == r { return true }
            let boundary = r.hasSuffix("/") ? r : r + "/"
            if candidate.hasPrefix(boundary) { return true }
        }
        return false
    }

    /// Test-facing: canonicalize the candidate and every root, then check containment.
    static func isContained(_ path: String, roots: [String]) -> Bool {
        isContainedCanonical(normalizeUser(canonical(path)),
                             canonicalRoots: roots.map { normalizeUser(canonical($0)) })
    }

    /// Directories that legitimately hold servable content: the parent folder of
    /// every library track, plus the agent's own cache (server-generated audio
    /// derivatives) and the Rekordbox share/db dir (cover art). Rebuilt from the
    /// (cached) parsed library on each call.
    static func allowedRoots() -> [String] {
        var roots = Set<String>()
        for track in RekordboxParser.shared.parse().tracks {
            let loc = track.location
            guard !loc.isEmpty else { continue }
            let dir = (loc as NSString).deletingLastPathComponent
            if !dir.isEmpty, dir != "/" { roots.insert(dir) }
        }
        // Agent-managed dirs: converted 16-bit/AIFF WAVs and extracted artwork.
        roots.insert(AppConfig.shared.cacheDir.path)
        let db = AppConfig.shared.dbPath
        if !db.isEmpty {
            let rbDir = (db as NSString).deletingLastPathComponent
            if !rbDir.isEmpty, rbDir != "/" {
                roots.insert(rbDir)
                roots.insert((rbDir as NSString).appendingPathComponent("share"))
            }
        }
        return Array(roots)
    }

    /// Canonicalized roots, memoized on a cheap library signature so the hot
    /// streaming path does not realpath every root on every (range) request — which
    /// on a large library over a slow external volume would add real latency.
    private static func canonicalAllowedRoots() -> [String] {
        let tracks = RekordboxParser.shared.parse().tracks
        let signature = "\(tracks.count)|\(tracks.first?.location ?? "")|\(tracks.last?.location ?? "")"
        rootsLock.lock(); defer { rootsLock.unlock() }
        if signature == cachedRootsSignature, !cachedCanonicalRoots.isEmpty {
            return cachedCanonicalRoots
        }
        let canon = allowedRoots().map { normalizeUser(canonical($0)) }
        cachedRootsSignature = signature
        cachedCanonicalRoots = canon
        return canon
    }

    /// The check the file endpoints call. Empty path or empty library ⇒ nothing is
    /// servable ⇒ deny.
    static func isAllowed(_ path: String) -> Bool {
        guard !path.isEmpty else { return false }
        return isContainedCanonical(normalizeUser(canonical(path)),
                                    canonicalRoots: canonicalAllowedRoots())
    }
}

// MARK: - Auto-update signature verification (6005: no unsigned RCE)

enum UpdateSignatureError: LocalizedError {
    case codesignFailed(String)
    case teamMismatch(String)
    case unsafePath(String)
    case missingSignature(String)
    case badSignature

    var errorDescription: String? {
        switch self {
        case .codesignFailed(let s):   return "Update signature invalid: \(s)"
        case .teamMismatch(let s):     return "Update signed by an untrusted team: \(s)"
        case .unsafePath(let s):       return "Refusing unsafe update path: \(s)"
        case .missingSignature(let s): return "Update signature file missing: \(s)"
        case .badSignature:            return "Update archive signature is invalid"
        }
    }
}

enum UpdateSignatureVerifier {
    /// The only Team ID whose Developer ID Application signature we accept for a
    /// self-update (see build_local_mac.sh — "team MM3Q8TJL85"). An update signed
    /// by anyone else, or unsigned, is refused BEFORE the running bundle is
    /// replaced — a compromised GitHub release cannot ship our signature.
    static let expectedTeamID = "MM3Q8TJL85"

    /// Detached-signature public key (contents of update_pub.pem, PKIX/SPKI). The
    /// matching private key lives ONLY in release CI (secret UPDATE_SIGNING_KEY) and
    /// signs `update.zip` → `update.zip.sig`. Same key on macOS and Windows, so one
    /// signature verifies identically everywhere. A compromised GitHub release can
    /// serve a zip but cannot produce a valid .sig without this private key.
    static let updatePublicKeyPEM = """
    -----BEGIN PUBLIC KEY-----
    MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE13+tB8Bq9f/Hi7ez4nff0roiRz4E
    LGxTQT37N0MMd+jwfqh4P+vmK4oMBsrvwhxkIRB60rgWcVbr8irPMYJLEw==
    -----END PUBLIC KEY-----
    """

    /// Verify an update archive against its detached ECDSA-P256 / SHA-256 (DER)
    /// signature using the baked update public key. Produced by
    /// `openssl dgst -sha256 -sign update_priv.pem`. Throws (fail-CLOSED) when the
    /// .sig is absent or does not verify — the caller MUST NOT apply on throw.
    static func verifyZipSignature(zipPath: String, sigPath: String) throws {
        try verifyZipSignature(zipPath: zipPath, sigPath: sigPath, publicKeyPEM: updatePublicKeyPEM)
    }

    /// Injectable-key overload for unit tests.
    static func verifyZipSignature(zipPath: String, sigPath: String, publicKeyPEM: String) throws {
        guard FileManager.default.fileExists(atPath: sigPath) else {
            throw UpdateSignatureError.missingSignature(sigPath)
        }
        let sigData = try Data(contentsOf: URL(fileURLWithPath: sigPath))
        guard let pubKey = try? P256.Signing.PublicKey(
                pemRepresentation: publicKeyPEM.trimmingCharacters(in: .whitespacesAndNewlines)),
              let signature = try? P256.Signing.ECDSASignature(derRepresentation: sigData) else {
            throw UpdateSignatureError.badSignature
        }
        let digest = try sha256OfFile(zipPath)
        guard pubKey.isValidSignature(signature, for: digest) else {
            throw UpdateSignatureError.badSignature
        }
    }

    private static func sha256OfFile(_ path: String) throws -> SHA256.Digest {
        guard let fh = FileHandle(forReadingAtPath: path) else {
            throw UpdateSignatureError.missingSignature(path)
        }
        defer { try? fh.close() }
        var hasher = SHA256()
        while case let chunk = fh.readData(ofLength: 1 << 20), !chunk.isEmpty {
            hasher.update(data: chunk)
        }
        return hasher.finalize()
    }

    static func teamIDMatches(_ actual: String?, expected: String) -> Bool {
        guard let actual = actual?.trimmingCharacters(in: .whitespacesAndNewlines),
              !actual.isEmpty else { return false }
        return actual == expected
    }

    /// Reject paths carrying shell metacharacters before they ever reach the
    /// osascript `do shell script` privileged-replace fallback (defense in depth
    /// against command injection through a crafted .app name inside the archive).
    static func isSafeShellPath(_ path: String) -> Bool {
        guard !path.isEmpty else { return false }
        let forbidden: Set<Character> = ["'", "\"", "`", "$", "\\", "\n", "\r",
                                         ";", "&", "|", "<", ">", "(", ")"]
        return !path.contains(where: { forbidden.contains($0) })
    }

    /// Extract `TeamIdentifier=` from `codesign -dv --verbose=4` output. Returns
    /// nil for an ad-hoc / unsigned bundle ("TeamIdentifier=not set").
    static func parseTeamID(fromCodesignOutput out: String) -> String? {
        for line in out.components(separatedBy: .newlines) {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.hasPrefix("TeamIdentifier=") {
                let v = String(trimmed.dropFirst("TeamIdentifier=".count))
                    .trimmingCharacters(in: .whitespaces)
                return (v.isEmpty || v == "not set") ? nil : v
            }
        }
        return nil
    }

    /// Verify a downloaded/extracted `.app`: structural signature integrity
    /// (codesign --verify --deep --strict) AND our Team ID. Throws (fail-closed)
    /// on any failure; the caller MUST NOT replace the running bundle on throw.
    static func verify(appPath: String) throws {
        let verify = runProcess("/usr/bin/codesign",
                                ["--verify", "--deep", "--strict", appPath])
        guard verify.status == 0 else {
            throw UpdateSignatureError.codesignFailed(
                verify.err.isEmpty ? "codesign --verify exited \(verify.status)" : verify.err)
        }
        let info = runProcess("/usr/bin/codesign", ["-dv", "--verbose=4", appPath])
        // codesign writes the signing info to stderr.
        let combined = info.err.isEmpty ? info.out : info.err
        let team = parseTeamID(fromCodesignOutput: combined)
        guard teamIDMatches(team, expected: expectedTeamID) else {
            throw UpdateSignatureError.teamMismatch(team ?? "(unsigned)")
        }
    }

    @discardableResult
    private static func runProcess(_ launchPath: String, _ args: [String])
        -> (status: Int32, out: String, err: String) {
        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: launchPath)
        proc.arguments = args
        let outPipe = Pipe(); let errPipe = Pipe()
        proc.standardOutput = outPipe
        proc.standardError = errPipe
        do { try proc.run() } catch { return (-1, "", "\(error)") }
        let outData = outPipe.fileHandleForReading.readDataToEndOfFile()
        let errData = errPipe.fileHandleForReading.readDataToEndOfFile()
        proc.waitUntilExit()
        return (proc.terminationStatus,
                String(data: outData, encoding: .utf8) ?? "",
                String(data: errData, encoding: .utf8) ?? "")
    }
}

// MARK: - Helper-binary download host pinning (6006)

enum ComponentHostPolicy {
    /// Hosts we accept helper-binary (cloudflared / ffmpeg / ffprobe) downloads
    /// from. The manifest is served by rimeo.app; the binaries live on rimeo.app
    /// or its GitHub-releases CDN. Pinning the host means a compromised manifest
    /// cannot redirect the download to an attacker server to drop an executable.
    static let allowedHosts: Set<String> = [
        "rimeo.app", "www.rimeo.app",
        "github.com",
        "objects.githubusercontent.com",
        "release-assets.githubusercontent.com",
    ]

    static func isAllowed(urlString: String) -> Bool {
        guard let url = URL(string: urlString),
              url.scheme?.lowercased() == "https",
              let host = url.host?.lowercased() else { return false }
        return allowedHosts.contains(host)
    }
}
